동기화 메모장 [STEP 1] 지스

hrjy6278 opened the pull request:
안녕하세요 비비!!! 🙌 완벽하지않은 스텝1 PR을 보냅니다! 저 혼자 끙끙 앓는것 보단 해결되지않는 점을 같이 보내는게 
더 수렁에 빠지지않을 것 같아 PR 보냅니다 ㅎㅎㅎㅎㅎㅎㅎ 
혼자 하는 첫 프로젝트여서 더 떨리기도 하네요 ㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎ
리뷰 잘 부탁 드리겠습니다!!!!!!!!!!!! @YebinKim 

## Step1 동작구현
- 아이폰 
 ![Simulator Screen Recording - iPhone 12 Pro Max - 2021-09-03 at 15 15 18](https://user-images.githubusercontent.com/71247008/131959918-5ee4b6d1-0b6d-4168-b3f1-60bfea1e2271.gif)

- 아이패드
 ![Simulator Screen Recording - iPad Pro (12 9-inch) (5th generation) - 2021-09-03 at 15 41 15](https://user-images.githubusercontent.com/71247008/131965879-1b95d867-b849-4ad4-ae72-828bb516118b.gif)

## 코드에 관하여
1. 처음에는 `TableViewDataSource`를 사용하여 구현했으나, 추후에 검색 기능을 위해 좀더 편리하고자 `DiffableDataSource` 를 사용했습니다. (그리고 한번도 써보질 못해서 이번기회에 공부하면서 쓴 이유도 있습니다.)

2. DataFormatter Locale을 기기에 맞는 지역에 맞게끔 구현하였습니다. 
```swift=
dateFormatter.locale = Locale(identifier: Locale.current.identifier)
```
3. iOS14 버전부터 사용가능한 UISplitViewController의 column-style layouts 방법으로 구현하였습니다.

4. 현재 기기 Width SizeClass 상태에따라 화면을 구현하였습니다.

5. 스택뷰를 활용하여 TableViewCell `Label`을  설정하였습니다.

## 코드에 관한 질문
1. 현재 DetailViewController에 TextView 가 메모 제목과, 메모 본문을 포함하고 있는데요! 단일뷰라서 제목과, 본문을 어떻게 둘 다 표시해줄까 생각하다가 현재 코드가 탄생하였습니다! 근데 너무 생각없이 짠 코드같아서 ㅎㅎㅎㅎ 더 좋은 방법이 있을지.. 궁금합니다! 
```swift=
memoContentsTextView.text.append(memo.title)
memoContentsTextView.text.append("\n\n")
memoContentsTextView.text.append(memo.body)
```

2. 하드코딩을 피하고자 enum을 활용하여 적용해보았는데요!..현재 활용방식이 어떤지에 대해 궁금합니다 ㅎㅎㅎㅎㅎ

3. 코드로 UI를 짜실때 비비의 팁이랄께 있을까요? 아지 초보라 스토리보드에서도 헤매는데 없이 하려니 더 힘들었습니다.. 지금도 에러가 나긴하는데 어떻게 생긴 에러인지 감도 못잡겠습니다...

4. 현재 UI를 구성하는 프로퍼티들은 클로저를 이용하여 초기화를 진행하였는데요! 비비는 어떻게 사용하시는지? 궁금합니다..!

5. `textView`  스크롤이 처음에 `텍스트뷰`의 중간에 계속 위치하는 에러가 있어 현재 아래코드 처럼 수정을 하니 제대로 되더라구요.. 근데 왜 저렇게 해야만 스크롤이 맨위에서 시작하는 건지 이해가 잘 안됩니다..
```swift=
memoContentsTextView.contentOffset = CGPoint(x: 0, y: 0)
```

6. 제가 `MemoListTableViewController`에서 TableViewController 를 상속받아 구현했는데요! 저는 TableView의 오토레이아웃을 지정해준적이 없는데 자동으로 꽉찬 화면으로 구현이 되더라구요..편리하기는 했으나, 이게 왜 그런건지 궁금해서 구글링도 해보았지만 답을 찾지 못했습니다.. 스토리보드에서 테이블뷰컨트롤러를 추가하면 기본적인게 셋팅되는 것 처럼 똑같은 원리 일까요??

## 해결되지 않은 문제

1. 현재 메모 list 화면에서 DetailViewController로 넘어가는 순간 디버그창에 에러 내용이 표시되며 작동이 됩니다. 사실 `TableViewDataSource` 프로토콜로 구현했을때는 이런 에러가 안났었는데, `DiffableDataSource`로 구현하니 에러가 나기 시작했습니다.. 에러내용을 봐도 감도 못잡겠고, 계속 PR이 지체될 것 같아 에러내용을 포함해 PR 보냅니다.. 

<details>
<summary> 에러내용 </summary>

```swift=
2021-09-02 20:11:28.629969+0900 CloudNotes[67460:9088556] [TableView] Warning once only: UITableView was told to layout its visible cells and other contents without being in the view hierarchy 
(the table view or one of its superviews has not been added to a window). This may cause bugs by forcing views inside the table view to load and perform layout without accurate information (e.g. table view bounds, trait collection, layout margins, safe area insets, etc), 
and will also cause unnecessary performance overhead due to extra layout passes. Make a symbolic breakpoint at UITableViewAlertForLayoutOutsideViewHierarchy to catch this in the debugger and see what caused this to occur, so you can avoid this action altogether if possible, or defer it until the table view has been added to a window. 
Table view: <UITableView: 0x7fa837024400; frame = (0 0; 0 0); clipsToBounds = YES; autoresize = W+H; gestureRecognizers = <NSArray: 0x600003dfbc90>; layer = <CALayer: 0x6000033d5380>; contentOffset: {0, 0}; contentSize: {0, 0}; adjustedContentInset: {0, 0, 0, 0}; 
dataSource: <_TtGC5UIKit29UITableViewDiffableDataSourceOC10CloudNotes27MemoListTableViewController7SectionVS1_4Memo_: 0x600003185b30>>
2021-09-02 20:11:28.645359+0900 CloudNotes[67460:9088556] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x6000010b9090 UIStackView:0x7fa835c1f7a0.leading == UILayoutGuide:0x600000a85340'UIViewSafeAreaLayoutGuide'.leading + 10   (active)>",
    "<NSLayoutConstraint:0x6000010b90e0 UIStackView:0x7fa835c1f7a0.trailing == UILayoutGuide:0x600000a85340'UIViewSafeAreaLayoutGuide'.trailing   (active)>",
    "<NSLayoutConstraint:0x6000010bdb30 'UISV-alignment' UILabel:0x7fa835c1eaa0.leading == UIStackView:0x7fa835c08e90.leading   (active)>",
    "<NSLayoutConstraint:0x6000010bdb80 'UISV-alignment' UILabel:0x7fa835c1eaa0.trailing == UIStackView:0x7fa835c08e90.trailing   (active)>",
    "<NSLayoutConstraint:0x6000010bd630 'UISV-canvas-connection' UIStackView:0x7fa835c08e90.leading == UILabel:0x7fa835c1f530.leading   (active)>",
    "<NSLayoutConstraint:0x6000010bd680 'UISV-canvas-connection' H:[UILabel:0x7fa835c1f2c0]-(0)-|   (active, names: '|':UIStackView:0x7fa835c08e90 )>",
    "<NSLayoutConstraint:0x6000010bda90 'UISV-canvas-connection' UIStackView:0x7fa835c1f7a0.leading == UILabel:0x7fa835c1eaa0.leading   (active)>",
    "<NSLayoutConstraint:0x6000010bdae0 'UISV-canvas-connection' H:[UILabel:0x7fa835c1eaa0]-(0)-|   (active, names: '|':UIStackView:0x7fa835c1f7a0 )>",
    "<NSLayoutConstraint:0x6000010bd7c0 'UISV-spacing' H:[UILabel:0x7fa835c1f530]-(>=30)-[UILabel:0x7fa835c1f2c0]   (active)>",
    "<NSLayoutConstraint:0x6000010bdc20 'UIView-Encapsulated-Layout-Width' UITableViewCellContentView:0x7fa835c1fe40.width == 27.3333   (active)>",
    "<NSLayoutConstraint:0x6000010b8dc0 'UIViewSafeAreaLayoutGuide-left' H:|-(0)-[UILayoutGuide:0x600000a85340'UIViewSafeAreaLayoutGuide'](LTR)   (active, names: '|':UITableViewCellContentView:0x7fa835c1fe40 )>",
    "<NSLayoutConstraint:0x6000010b8e60 'UIViewSafeAreaLayoutGuide-right' H:[UILayoutGuide:0x600000a85340'UIViewSafeAreaLayoutGuide']-(0)-|(LTR)   (active, names: '|':UITableViewCellContentView:0x7fa835c1fe40 )>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x6000010bd7c0 'UISV-spacing' H:[UILabel:0x7fa835c1f530]-(>=30)-[UILabel:0x7fa835c1f2c0]   (active)>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging
```

</details>
<br>


2. 아이폰의 `Width` 가 `Regular` 일 때 스플릿뷰의 컬럼부분이 좌측이 너무 횡 하고 띄어져 있습니다. 이 부분을 해결을 하지 못하였습니다. ![Simulator Screen Shot - iPhone 12 Pro Max - 2021-09-01 at 22 30 18](https://user-images.githubusercontent.com/71247008/131963776-165618ea-4ed0-41c9-9b79-4f4ce5396aba.png)
 
 
 3. 이거는 시뮬레이터 문제인지 제가 구현을 잘못한건지는 모르겠지만 지역에 맞는 날짜형식을 지정하는 부분에서 `Locale.current.identifire` 가 `en_KR` 로 계속 출력이 되더라구요. 이래서 날짜형식이 계속 영어권 형식으로 포맷팅되어 나옵니다. 제가 틀린 부분이 있는 건지 잘 모르겠습니다.😢
 

4. 이 부분은 영상을 보시는게 더 좋으실 것 같습니다. 네비게이션 컨트롤러에 계속 화면이 push 되어 쌓이고 있습니다..😵‍💫
![Simulator Screen Recording - iPad Pro (12 9-inch) (5th generation) - 2021-09-03 at 16 11 26](https://user-images.githubusercontent.com/71247008/131966577-7928382a-32b6-4ed3-ad1c-266d6a5205f2.gif)

hrjy6278 committed:
feat: SwiftLint 라이브러리 적용

- swiftlint.yml 을 생성하여 swiftlint 에 제외시킬 파일을 추가하였다.

hrjy6278 committed:
feat: Main 스토리보드 삭제

- SceneDelegate에서 윈도우에 rootViewController를 SplitViewController로 지정하였다.

hrjy6278 committed:
chore: Test Data 폴더 생성 및 MVC 폴더 생성

- TestData 폴더에 sample.json 파일을 추가하였다.
- 기존 ViewController 네이밍을 SplitViewController 로 수정하였다.
- Info.plist에서 main 스토리보드를 삭제하였다.

hrjy6278 committed:
feat: MemoListViewController,MemoDatailViewController 클래스 생성

- 스플릿뷰에 보여주기위한 MemoListViewController, MemoDatailViewController를 구현하였다.
- 스플릿뷰컨트롤러에 두개의 컨트롤러를 셋업하는 setupChildViewController 를 구현하였다.
- 스플릿뷰의 기본 디스플레이 모드를 설정하는 setupSplitViewDisPlayMode 메서드를 구현하였다.
- sample Json 파일을 파싱하기위한 JsonManager와 Memo 타입을 구현하였다.

hrjy6278 committed:
feat: MemoListTableViewController 의 rightBarButtonItem 설정

hrjy6278 committed:
feat: MemoListTableViewCell 클래스 구현

- 테이블뷰에 보여질 Cell 클래스를 구현하였다.
- Cell안에 들어갈 Label들을 프로퍼티로 구현하였다.

hrjy6278 committed:
feat: MemoListTableViewCell 의 configureLabel 메소드 구현(ING)

- Cell에 보여질 body와, lastModified 를 StackView로 처리하는 메서드 구현중
- MemoListTableViewController 에 Cell을 Register 하는 메서드 구

hrjy6278 committed:
feat: MemoListTableViewCell 의 configureDescriptionStackView 메서드 구현완료

- 메모의 설명과와, 날짜정보를 알려주는 Label을 스택뷰를 활용하여 configureDescriptionStackView 메서드로 구현하였다.
- 하드코딩을 지양하기 위하여 NameSpace 열거형 구현.

hrjy6278 committed:
feat: TableView Cell의 label들을 소유한 containerStackView 구현

- 기존 descriptionStackView의 Constant는 삭제하였다. 대신 containerStackView에 add가 되었다.

hrjy6278 committed:
feat: 날짜형식을 기기 지역에 맞는 날짜형식으로 바꿔주는 기능 구현

hrjy6278 committed:
feat: SplitViewController에 compact 모드일때 보여질 ViewController 추가

hrjy6278 committed:
feat: DateilViewController의 BarButtonItem 구현

hrjy6278 committed:
feat: MemoDatailViewController 에서 메모내용을 보여줄 TextView 구현

hrjy6278 committed:
feat: tableView(_:didSelectRowAt:) 메서드 구현

- 스플릿뷰컨트롤러의 컴팩트 뷰컨트롤러를 삭제하였다.
- tableView(_:didSelectRowAt:) 메서드가 호출될때 사이즈에 맞게 뷰를 보여주게 하였다.

hrjy6278 committed:
refactor: sample.json 파싱을 위한 로직 수정

-sample.json은 샘플파일이기때문에, SampleMemo 타입을 만들어 처리하도록 구현하였다.

hrjy6278 committed:
style: 주석 및 접근제어자 추가

hrjy6278 committed:
style: 코드컨벤션 반영

hrjy6278 committed:
refactor: Diffable Data Source 적용

- 기존 TableViewDataSource를 Diffable Data Source로 교체하였다.

hrjy6278 committed:
refactor: 기기 width compact 일시 TextView color 변경 로직 추가

-SlitViewController 첫 secondary VC를 UIViewController로 변경
- 전반적 접근제어자 수정

hrjy6278 committed:
docs: step1 UML readme 추가

YebinKim commented on CloudNotes/CloudNotes/Model/Memo.swift:
```diff
@@ -0,0 +1,59 @@
+
+import Foundation
+
+struct Memo: Decodable, Hashable {
```
Hashable 프로토콜은 어떤 의미로 채택하고자 하셨을까요? 🤔

hrjy6278 replied:
Diffable DataSource가 제네릭타입인데 조건이 `Hashable`을 만족하는 타입이여야 된다고 명시가 되어 있었습니다.(마치 딕셔너리의 키와 값처럼)
`Hashable` 프로토콜을 채택하면 은 해당 타입이 `hash`가 가능하다는 뜻이며 `hash`는 그 자체로 유일하게 표현될 수 있는 값이라고 알고 있습니다.

YebinKim replied:
오호 상세한 설명 고맙습니다 ㅎㅎ
DiffableDataSource에서 사용하기 위해 Hashable 프로토콜의 채택이 필요했겠군요~!

YebinKim commented on CloudNotes/CloudNotes/Controller/MemoDatailViewController.swift:
```diff
@@ -0,0 +1,73 @@
+    }
+    
+    deinit {
+        print("\(self) 제거되었음")
```
deinit 확인하는 습관 멋지네요! 😆
print 메서드 대신 스위프트에서 기본으로 제공하는 로그 타입이나 debugPrint(_:separator:terminator:) 을 활용해보는 것 또한 생각해보면 좋을 거 같아요~ (가능하면 머지 커밋에는 print 메서드를 남기지 않는 것이 좋아요)

hrjy6278 replied:
앗 넵 `debugPrint(_:separator:terminator:) ` 을 활용하여 보겠습니다 !

YebinKim commented on CloudNotes/CloudNotes/Controller/MemoDatailViewController.swift:
```diff
@@ -0,0 +1,73 @@
+        memoContentsTextView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor).isActive = true
+        memoContentsTextView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor).isActive = true
+        memoContentsTextView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor).isActive = true
+        memoContentsTextView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor).isActive = true
```
[NSLayoutConstraint.activate(_:)](https://developer.apple.com/documentation/uikit/nslayoutconstraint/1526955-activate) 메서드로도 오토레이아웃을 설정할 수 있어요~
라인마다 isActive = true 를 해주는 것이 번거롭다면 이 방법 또한 생각해볼 수 있을 거 같습니다 ☺️

YebinKim commented on CloudNotes/CloudNotes/Controller/MemoListTableViewController.swift:
```diff
@@ -0,0 +1,105 @@
+
+// MARK: - Data Source
+extension MemoListTableViewController {
+   private enum Section {
```
일부 라인의 탭 간격이 3 whitespaces 여서 다른 코드들과 간격이 다른 것으로 보이는데요 🤔
정렬이 필요한 코드는 드래그하여 `control + i` 단축키를 누르면 일괄적으로 탭 간격을 맞출 수 있습니다~!

<img width="332" alt="whitespaces" src="https://user-images.githubusercontent.com/38066039/132388190-a5b602d8-53d9-43d5-9421-4d81d5f99d97.png">

YebinKim commented on CloudNotes/CloudNotes/Controller/MemoListTableViewController.swift:
```diff
@@ -0,0 +1,105 @@
+            guard let cell = tableView.dequeueReusableCell(withIdentifier: MemoListTableViewCell.identifier, for: indexPath) as? MemoListTableViewCell else { fatalError() }
+            
+            cell.accessoryType = .disclosureIndicator
+            cell.configure(with: memo)
```
cell 의 UI를 configure 메서드를 통해 설정함으로써, UI를 그리는 역할에 대한 책임을 cell 이 가지도록 작성하셨네요! 🤩👏

YebinKim commented on CloudNotes/CloudNotes/Controller/MemoListTableViewController.swift:
```diff
@@ -0,0 +1,105 @@
+        })
+    }
+    
+   private func snapshot(ofMemo memo: [Memo]) {
```
배열을 매개변수로서 전달받는다면 단수형의 명사보다는 복수형의 명사를 사용하면 더 이해하기 쉬울 거 같아요~

hrjy6278 replied:
이 부분은 수정이 필요하겠네요..! 감사합니다!

YebinKim commented on CloudNotes/CloudNotes/View/MemoListTableViewCell.swift:
```diff
@@ -0,0 +1,108 @@
+
+import UIKit
+
+class MemoListTableViewCell: UITableViewCell {
```
⭐️ MemoListTableViewCell 이 재사용될 때는 UI 초기화가 필요하지 않을까요? 🤔

hrjy6278 replied:
아 이 부분도 미쳐 고려하지 못했습니다 ㅜ 수정하겠습니다~!!!

YebinKim commented on README.md:
```diff
@@ -1,7 +1,5 @@
-- 이 저장소를 자신의 저장소로 fork하여 프로젝트를 진행합니다
-- 자신의 브랜치에 PR을 보내는지 꼭 확인한 후 PR을 보냅니다
+### 동기화 메모장 Step1 UML
+![CloudMemoUML](https://user-images.githubusercontent.com/71247008/131958423-83f991e2-c495-43e7-a74b-2803b198c938.png)
```
깔끔한 UML 다이어그램 덕분에 프로젝트 구조를 파악하는데 수월했습니다~! 😆👍

YebinKim reviewed (changes requested):
지스 안녕하세요~!
3주간 동기화 메모장 프로젝트를 리뷰할 비비입니다 😄

스토리보드 없이 코드로만 레이아웃을 구성하고,
`DiffableDataSource` 를 사용해 테이블뷰를 구성하고 👏👍
전반적인 코드의 배치와 네이밍도 깔끔하네요!

코드에서 고민을 많이 하셨음이 느껴졌습니다 😊
코멘트로 전달주신 문제점들과 같이 동작 상 오류가 있는 부분도 일부 보이는데요.
관련하여 상세한 라인별 코멘트는 아래에서 드리고, 우선 질문부터 답변 드려볼게요-!



---

## 코드에 관한 질문

### 1. `DetailViewController` 에서 `textview.text` 를 설정하는 방법에 관하여

현재 동작 상으로는 문제가 있어보이진 않는데요!
한 가지 우려사항이 있다면 **`memoContentsTextView.text` 를 초기화**해주는 코드가 없기 때문에 만약 `memoContentsTextView.text` 에 다른 문자열이 있다면 예상과 다른 화면을 보게 될 수 있을 거 같아요.
그래서 저라면 `memoContentsTextView.text` 에 매번 각각의 문자열을 append 하는 대신 완성된 문자열을 대입하는 방식으로 코드를 작성해볼 거 같습니다 😄

```swift
let appendedText: String = memo.title + "\n\n" + memo.body
memoContentsTextView.text = appendedText
```





### 2. enum을 활용하는 방식에 관하여 

하드코딩을 줄이고자 enum을 사용하는 것은 좋은 습관이라고 생각합니다! 😆👍
그리고 현재의 enum들도 각각의 용도와 필요에 따라 적절한 네이밍, 접근제어자로 작성하셨다고 생각해요~

`DiffableDataSource` 에 필요한 SectionType 으로서의 사용을 위해서도 enum을 사용한 것은 적절한 선택으로 보입니다!





### 3. 코드로 UI를 작성할 때의 팁 🍯

코드로 UI를 작성하면서 특히 어려운 점이 있다면.. 오토레이아웃을 적용할 때 여러 상황에서 오토레이아웃이 유효한지를 (런타임 중 에러가 발생하지 않는지를) 파악하기 어려운 점이라고 생각해요~!
그래서 저는 임시 프로젝트를 하나 만들어서 스토리보드 위에 제가 적용하고자 하는 오토레이아웃을 동일하게 적용해봅니다 ㅎㅎ

회전을 지원할 경우 **세로 방향에서는 오토레이아웃이 유효하지만 가로 방향에서는 깨지는 경우**도 있고,
**화면 사이즈에 따라 오토레이아웃이 깨지는 경우**도 있기 때문에 최소 사이즈, 최대 사이즈,  regular x compact 사이즈, regular x regular 사이즈 등 다양한 사이즈 클래스에서 확인해보는 것도 필요해요!

현재 발생한 에러에 대해서는 아래의 해결되지 않은 문제 1번 질문을 통해 코멘트 드릴게요-! 😊





### 4. 코드로 UI를 작성하는 방법에 관하여

저도 현재 지스가 작성하신 것과 같이 클로저를 이용해 초기화하는 것을 선호합니다!
작성하는 방법은 여러가지가 있고 정답은 없지만,
저는 UI의 기본 옵션과 관련된 설정 및 변경되지 않는 설정은 클로저를 통해 수행하고, 오토레이아웃 또는 업데이트가 필요한 설정은 따로 메서드를 통해 수행할 수 있도록 구분하는 편이에요~





### 5. `textView` 의 스크롤이 중간에 위치하는 에러에 관하여

음.. 저도 정확한 문제 원인을 파악한 것은 아니지만, 추측해보자면 **`showContents(of:)` 메서드에서 텍스트를 추가할 때 레이아웃 업데이트 사이클이 제대로 수행되지 않아서** 그런 걸로 보이는데요 🤔
(레이아웃 업데이트 사이클에 대해서는 `viewWillLayoutSubviews()` 메서드와 `viewDidLayoutSubviews()` 메서드가 각각 언제 호출되는지, 어떤 역할을 하는지, 뷰의 레이아웃을 업데이트하는 메서드는 무엇이 있는지 정도만 이해하고 넘어가도 충분합니다!)

이를 해결하기 위해서는 3가지 방법을 생각해볼 수 있을 거 같아요!

1. 텍스트를 추가한 후 레이아웃 업데이트 사이클을 강제로 호출시키거나 -> (`showContents(of:)` 메서드에서 text를 append한 후 memoContentsTextView 에 레이아웃 업데이트 메서드를 호출)
2. UI 업데이트가 메인 스레드 안에서 동작할 수 있도록 하거나
3. 현재와 같이 뷰가 나타나기 전에 contentOffset 을 0으로 지정하는 방법 -> (현재 코드는 `viewDidLoad()` 에서 contentOffset을 지정하고 있지만 좀 더 정확한 위치는  `viewWillAppear(animated:)` 로 생각됩니다 ☺️)

이 중에서 또는 다른 방법으로 해결해볼 수 있을텐데요.
어떤 방법이 적절할지는 지스가 결정해주시면 됩니다! 🙌





### 6. UITableViewController 에서 tableview 의 레이아웃이 지정되는 방법에 관하여

오호! 저도 지스의 질문을 보고 궁금증이 들어서 찾아보았는데요 ㅎㅎ
애플이 이 부분에 대해 명확히 공개하고 있진 않지만, 아마 **UITableViewController의 `viewWillLayoutSubviews()` 메서드 이전에 tableview의 frame을 지정하는 코드가 호출**되고 있는 거 같아요.

우선 UITableViewController의 주석부터 살펴보자면

```
Creates a table view with the correct dimensions and autoresizing, setting the datasource and delegate to self.
```

테이블뷰를 알아서 잘 그려준다고 하고 있지만..
우리는 이걸 직접 확인해보고 싶으니 여러가지 테스트를 해보았는데요!
원인을 시원하게 밝혀내진 못 했지만 결론을 정리하자면 아래와 같습니다.

1. `viewWillAppear(animated:)` 에서 tableview의 frame을 변경하면 영향이 없음
2. `viewWillLayoutSubviews()` 에서 tableview의 frame을 변경하면 화면 첫 진입 시 적용되며, 회전 시마다 레이아웃이 적용됨
3. `viewDidLayoutSubviews()` 에서 tableview의 frame을 변경하면 화면 첫 진입 시 적용되며, 회전 시마다 레이아웃이 적용됨
4. `viewDidAppear(animated:)` 에서 tableview의 frame을 변경하면 화면 첫 진입 시 적용되며, 회전 시에는 레이아웃이 적용되지 않음 (당연합니다 회전 시에는 viewDidAppear가 불리지 않으니까요)
5. `viewWillTransition(to:with:)` 에서 tableview의 frame을 변경하면 화면 첫 진입 시 적용되지 않으며(당연합니다 화면 첫 진입 시에는 호출되지 않으니까요), 회전 시에는 레이아웃이 적용되었다가 다시 화면에 꽉 차는 애니메이션이 보임
6. tableview의 오토레이아웃 적용을 해제해도 동작에는 영향이 없음
7. tableview의 오토리사이징마스크 적용을 해제해도 동작에는 영향이 없음
8. 각 오버라이딩 메서드에서 super 메서드의 호출을 제거해도 동작에는 영향이 없음 (그래서 아예 다른 메서드를 통해 tableview의 frame이 변경되고 있는 거 같아요)
9. tableview의 delegate와 dataSource를 지정하지 않아도 레이아웃을 그리는데는 영향이 없음 🤔

즉, 레이아웃이 업데이트 될 때마다 `viewWillLayoutSubviews()` 이전에 frame을 지정하는 코드가 있기 때문에 화면에 꽉 차게 보이는 것이 아닐까.. 추측해보았습니다.

답변 내용이 다소 복잡하고 원인 또한 시원하게 밝혀내진 못 했지만 어떤 과정으로 이런 결과를 도출했는지는 회의실에서 보여드릴게요! 😆



---

## 해결되지 않은 문제

### 1. 레이아웃 에러가 발생하는 문제

우선 에러 내용부터 살펴보자면 

``````
UITableView was told to layout its visible cells and other contents without being in the view hierarchy
``````

UITableView가 화면에 보이기 전에 (윈도우의 뷰 계층에 들어오기 전에) 레이아웃을 적용하려고 하고 있다 (dataSource.apply 를 호출하고 있다) 로 해석해볼 수 있는데요.
더 직관적으로 얘기해보자면 dataSource.apply 를 최초 호출하는 시점에 뷰가 없기 때문에 에러를 표시해주고 있어요.
그래서 뷰가 완전히 로딩된 시점에 UI 업데이트가 동작할 수 있도록 **메인 스레드에서 dataSource.apply 를 호출**해줌으로써 해결하는 방법이 있어요 😄

``````swift
DispatchQueue.main.async {
    self.dataSource?.apply(snapShot, animatingDifferences: false, completion: nil)
}
``````

그리고 하나의 방법이 더 있는데요!
진리의 [공식문서](https://developer.apple.com/documentation/uikit/uitableviewdiffabledatasource/3375811-apply)를 참고해보자면 아래와 같이 Discussion에 적혀있어요.

> **Discussion**
> ~ 생략 ~
> You can safely call this method from a background queue, but you must do so consistently in your app. Always call this method exclusively from the main queue or from a background queue.

dataSource.apply 메서드는 백그라운드 스레드든 메인 스레드든 항상 **일관된 큐에서 호출하는 것이 필요하다**고 언급하고 있는데요.
즉, 일관된 큐에서 동작하기만 한다면 백그라운드 스레드에서도 안전한 호출을 보장할 수 있기 때문에 에러가 발생하지 않을 수 있는거죠 😯
그래서 DispatchQueue 를 통해 백그라운드 스레드에서 실행시켰을 때도 문제를 해결할 수 있어요!

``````swift
DispatchQueue.global().async {
    self.dataSource?.apply(snapShot, animatingDifferences: false, completion: nil)
}
``````





### 2. 아이폰의 `Width` 가 `Regular` 일 때 셀의 왼쪽이 비어보이는 문제

앗 이건 노치가 있는 디바이스에서 콘텐츠가 잘리지 않도록 SafeArea가 적용된 것으로 보이는데요 🤔
저는 이 부분이 오류라고 생각되지는 않지만.. 빈 영역을 제거하고자 한다면 tableview의 [insetsContentViewsToSafeArea](https://developer.apple.com/documentation/uikit/uitableview/2921665-insetscontentviewstosafearea) 플래그를 false로 설정함으로써 SafeArea의 적용을 해제할 수 있을 거에요~





### ⭐️ 3. `Locale.current.identifier` 가 올바르게 포맷팅되지 않는 문제

오호 저도 이 문제는 처음 겪어보았는데요!
`Locale.current.identifier` 는 프로젝트의 다국어로서 지원되는 언어에 대해서만 정보를 가져올 수 있는 것으로 보여요 🤔
문서를 봤을 때는 사용자의 지역 설정에 따라 정보를 가져와야 할 것 같은데.. 실기기에서나 시뮬레이터에서나 지역 설정을 바꿔봐도 `Locale.current.identifier` 는 바뀌지 않네요 😅
하지만, 프로젝트에 다국어를 추가하면 en_ko 로 출력이 되는걸 확인할 수 있어요!

그리고 공식문서를 봤을 때, 설정앱에서의 사용자의 설정 언어를 가져오기 위한 방법으로는 [preferredLanguages](https://developer.apple.com/documentation/foundation/locale/2293155-preferredlanguages) 를 사용하는 것을 권장하고 있으니 이 프로퍼티를 활용해보는 것 또한 생각해볼 수 있을 거 같아요~





### ⭐️ 4. 네비케이션 컨트롤러에 화면이 계속 push되는 문제

음.. 우선 화면이 계속 push되는 이유는 코드상으로 계속 네비게이션 컨트롤러에 push를 수행하고 있기 때문이에요!
현재 `SplitViewController`의 secondary 는 셀의 선택이 있을 때마다 `showDetailViewController` 메서드가 호출되고 있기 때문에 계속 네비게이션 컨트롤러에 `MemoDatailViewController` 가 쌓이고 있는 구조에요.
문제를 해결하기 위한 해결책도 제시할 수 있겠지만.. 그 이전에, **`SplitViewController`의 secondary 는 화면을 push 시킴으로써 동작해야하는 것이 맞는지**를 먼저 고민해보는 게 좋을 거 같아요 ☺️

이 문제에 대한 코멘트는 직접 고민해보실 수 있도록 간략하게 작성했는데요.
어떤 맥락으로 코멘트를 드린 것인지 의문이 든다면 좀 더 상세하게 설명드릴 수 있도록 할게요~!

그리고 동작 오류가 하나 더 있어서 공유드려요~ (위와 같이 화면의 전환 방식을 고민해보면 이 오류 또한 해결될 수 있을 거 같아요)
상세 재현 동선은 아래와 같습니다. 😊
- 가로가 compact인 상태 > MemoDatailViewController 진입 (임의의 셀 선택) > 가로 회전 > secondary 가 아닌 primary에 MemoDatailViewController 가 푸시된 모습 확인 됨

![SplitViewController](https://user-images.githubusercontent.com/38066039/132392351-50f5bb35-ad9b-4c29-b92e-0f511c7bdc3f.png)



---

코멘트가 많이 길어졌는데요..!
문제 파악에 시간이 걸리기도 했지만 최대한 객관적인 정보를 바탕으로 설명을 풀어내느라 시간이 많이 소요되기도 했네요 😂
각 질문에 대한 답변과 라인별 코멘트는 천천히 확인해보신 후 궁금한 점이 있다면 언제든 DM 또는 깃헙 코멘트 주세요~
다음 PR부터는 코멘트도 좀 더 빠르게 드릴 수 있도록 할게요.

**이번 스텝에서의 핵심경험을 채우기 위해 필수로 확인해야 할 코멘트는 ⭐️ 로 표시해두었습니다 ☺️**
특히 네비케이션 컨트롤러에 화면이 계속 push되는 문제는 다음 스텝 수행 전 수정을 진행하는 것이 필요할 거 같아 Request Changes를 통해 Step1을 한 번 수정하는 시간을 가져보았으면 해요-!



+)
깃 레포에 누락된 파일이 있는 거 같아요!
TestData 의 경로가 어떻게 설정되어 있는지 확인 한 번 부탁드려요 😊

<img width="151" alt="TestData" src="https://user-images.githubusercontent.com/38066039/132392321-715a2155-6e58-4a8e-a38f-77a1f92632ef.png">

hrjy6278 committed:
fix: 날짜와 시간을 지역화에 맞게끔 버그 수정

- 날짜시간 지역화 해주는 로직을 재사용성을 위해 연산프로퍼티가 아닌, extension DateForamatter로 적용하였다.

hrjy6278 committed:
fix: Step1 피드백 반영

- snapshot.apply를 비동기적으로 처리되도록 함.
- 누락되었던 sample.json 파일 추가
- DetailViewContoller의 메모 표시 로직 수
- MemoListTableViewController 의 prepareForReuse() 메서드 추가

hrjy6278 committed:
style: UITableViewDiffableDataSource 클로저를 후행클로저로 적용

hrjy6278 committed:
fix: MemoDetailVC에서 shocontents메서드가 호출되었을때 레이아웃을 즉시반영하도록 수정

- Layout isActive 부분을 NSLayoutConstraint.activate 로 변경

hrjy6278 committed:
refactor: SplitView 화면 전환방식 변경

- 기존 TableViewCell을 탭하면 ListTableVC에서 DetailVC를 호출하는 방식이 아닌, SplitVC에서 호출하도록함.
- 기존 DetailVC 호출하던 로직을 showDetail 메서드가 아닌 show 메서드를 사용하도록 변경
- 현재 화면 방식에 따른 TextView Background Color 로직을 ViewWillAppear 로 옮김.

hrjy6278 committed:
refactor: viewWillAppear에 실행되던 congifureTextViewBackGroundColor() 위치 변경

- 화면전환이 될 때마다 판단 할 수 있도록 traitCollectionDidChange(_:) 메서드로 위치를 변경하였다.

YebinKim commented on CloudNotes/CloudNotes/Utility/Protocol/MemoListDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol MemoListDelegate: AnyObject {
```
Delegate 패턴을 사용해 MemoListViewController 화면에서의 Cell 선택 동작에 대한 책임을 위임시키셨군요!
테이블뷰의 멋진 활용입니다 🤩

YebinKim commented on CloudNotes/CloudNotes/View/MemoListTableViewCell.swift:
```diff
@@ -0,0 +1,116 @@
+        lastModifiedLabel.text = DateFormatter.localizedString(of: memoItem.lastModified)
+    }
+    
+    override func prepareForReuse() {
+        titleLabel.text = nil
+        bodyLabel.text = nil
+        lastModifiedLabel.text = nil
+    }
```
셀 재사용 시의 초기화를 추가하셨군요! 👀👍

YebinKim commented on CloudNotes/CloudNotes/View/MemoListTableViewCell.swift:
```diff
@@ -0,0 +1,116 @@
+        
+        NSLayoutConstraint.activate([
+            containerStackView.topAnchor.constraint(equalTo: contentView.safeAreaLayoutGuide.topAnchor),
+            containerStackView.leadingAnchor.constraint(equalTo: contentView.safeAreaLayoutGuide.leadingAnchor),
```
세로 화면에서 레이블이 화면 왼쪽 가장자리에 딱 붙어있어 가독성이 떨어지는 거 같아요 🤔
혹시 의도된 디자인이었을까요?

![Layout](https://user-images.githubusercontent.com/38066039/132508575-022533b3-bd79-44e1-846b-7d4bac686fa1.png)

YebinKim reviewed (approved):
지스 안녕하세요~
빠르게 코멘트 반영해주시느라 고생 많으셨습니다!! 🥳👏
PR 통해서 전달주셨던 문제가 모두 깔끔하게 수정되었네요 👍

DM 통해서 전달드렸듯, 이번 스텝에서의 핵심 경험은 모두 충족되었다고 생각해 이대로 머지 진행할게요~
추가 코멘트 및 git repo 에 불필요한 파일이 올라간 건에 대해서는 이후 스텝에서 확인 부탁드립니다 🙏

Step1 진행 고생 많으셨어요-!
